### PR TITLE
works: Add RSS feed

### PR DESCRIPTION
**Pull request by novalis**
I haven't run the tests because the tests don't seem to work on my vagrant image (and they take a really long time to not work).  Hopefully, Travis will run and pass.  Manual testing seems to work, though.

### CODE REVIEW

File: app/controllers/works_controller.rb [hunk 1]
```diff
@@ -5,11 +5,11 @@ class WorksController < ApplicationController
   # only registered users and NOT admin should be able to create new works
   before_filter :load_collection
   before_filter :load_owner, :only => [ :index ]
-  before_filter :users_only, :except => [ :index, :show, :navigate, :search, :collected, :edit_tags, :update_tags ]
-  before_filter :check_user_status, :except => [ :index, :show, :navigate, :search, :collected ]
+  before_filter :users_only, :except => [ :index, :feed, :show, :navigate, :search, :collected, :edit_tags, :update_tags ]
```
Comment (novalis) [inline app/controllers/works_controller.rb]: This was like this when I found it, so I'm not going to change it unless I have to.

The rest (and the test failure), I'll fix.

**Comment by ariana-paris** [comment]
Hello,

Thank you for your contribution. I see you've already met our faithful Hound and his many objections to our sometimes ageing code. :)

Can you give us more details on what your pull request is designed to do? It looks as though it's adding an RSS feed per work; bearing in mind that we will soon be hitting 2 million works, we're not convinced this is a good idea, but perhaps you could let us know a bit more.

If you're interested in further submissions, we have a [backlog of approved issues](https://code.google.com/p/otwarchive/issues/list) (though get in touch by commenting on the issue before starting work as some may need clarification)

Cheers,

Ariana
Archive development team

**Comment by novalis** [comment]
I would like an RSS feed per work because otherwise I have to manually reload http://archiveofourown.org/works/3659997/navigate all the time, which is surely more server-intensive than generating an RSS feed.  Also, most works are probably already complete (or abandoned), so most RSS feeds will probably be unused.

My next-best alternative is to write a screen-scraper to scrape those navigate pages and generate my own RSS feed.  I guess I could also "subscribe", but I want to learn about updates when I'm in the mood to read things, rather than whenever I happen to check my mail.

In other words, I'm scratching my own itch here.

I haven't investigated your caching setup, but it would of course be possible to cache RSS feeds and invalidate the cache on work/chapter updates.  Would it be helpful for me to add that?

**Comment by elzj** [comment]
Thanks for your pull request! We tend to be a little conservative about adding new features nowadays, because we want to keep the overhead of testing and maintaining the code to a level that our team of volunteers can keep up with, which gets tough as things get more complex and the codebase gets older. In this case, what looks like a simple addition has a few more hidden gotchas in addition to the unrevealed works issue, although the pull request looks like a good start:

1) Draft works (where work.posted == false) shouldn't have RSS feeds.

2) Draft chapters in posted works shouldn't appear in the feed until the chapters themselves are posted.

3) If it's a feed, I'm guessing you would want the newest chapter updates first so as not to confuse readers. But it might get complicated, because creators can reorder chapters, so the newest one might theoretically not be the last by position. That might require some thought as to what the best behavior is.

4) Restricted works that aren't publicly accessible shouldn't have public RSS feeds.

5) The "revised at" field on works is what we use as the public updated date, since "updated at" might be affected by background changes or other kinds of cache busting. And then for chapters, you'd probably want the "published at" date to give some context as to when the chapter was posted. (Our chapter and work dating systems really need an overhaul.)

6) We'd definitely want caching on feeds like this, and then those should expire appropriately when there are new chapters or changes to existing chapters.

7) There are already a painful number of links in our work headers, and I'm not sure we want to add more before we make some design changes to consolidate those.

The number of things to test and consider starts to add up (8 year old systems! \o/ /o\), and we do already have a subscription feature for this purpose, so we'll probably want to discuss whether this is something we want to adopt. It might also make sense to change the existing tag-based feeds to reflect updates and not just new works, which might meet your needs for non-email-based tracking. We can let you know after we get a chance to discuss this internally, but definitely wanted to say thanks for your interest and for stopping by to contribute!

Elz
Archive development team

**Comment by novalis** [comment]
> 1) Draft works (where work.posted == false) shouldn't have RSS feeds.

Fixed.  For some reason, I assumed that the before_filters would take care of this, but I would have had to change it anyway due to caching.

> 2) Draft chapters in posted works shouldn't appear in the feed until the chapters themselves are posted.

Right now, you only see those chapters if you're the author or an admin. I guess I'll go ahead and change that, just so that caching is easier.

> 3) If it's a feed, I'm guessing you would want the newest chapter updates first so as not to confuse readers. But it might get complicated, because creators can reorder chapters, so the newest one might theoretically not be the last by position. That might require some thought as to what the best behavior is.

Typically (I only tested NewsBlur, but this is my assumption), RSS feeds end up ordered by date, regardless of the order in the file.  I'll change it anyway, since it can't hurt.  I believe that the feed order (since the feed is truncated at ten entries) and feed date should be chapter initial publication time, since what RSS users want is a way to be notified of new content.  This requires a little db migration (we can fall back to position order, in case we end up with two published at once).

4) Restricted works that aren't publicly accessible shouldn't have public RSS feeds.

I have decided to implement this by having no RSS feeds at all for restricted works.  This is because Rails 3, apparently, doesn't have an easy way to do conditional caching.  

> 5) The "revised at" field on works is what we use as the public updated date, since "updated at" might be affected by background changes or other kinds of cache busting. And then for chapters, you'd probably want the "published at" date to give some context as to when the chapter was posted. (Our chapter and work dating systems really need an overhaul.)

Fixed.

> 6) We'd definitely want caching on feeds like this, and then those should expire appropriately when there are new chapters or changes to existing chapters.

Fortunately, it appears that we already have most of the infrastructure necessary to do this.  Done.

Also, while I was in there I noticed that feed_sweeper expires caches even when updates affect only unpublished chapters.  Want me to fix that (in a separate PR, of course)?

> 7) There are already a painful number of links in our work headers, and I'm not sure we want to add more before we make some design changes to consolidate those.

I have not yet fixed this.  Here's my proposal: what if we just added a link tag to the <head> like this:
&lt;link rel="alternate" type="application/rss+xml" title="Feed" href=".../feed.atom" /&gt;

I think most feed readers will be able to pick up the feed from that, and I could document it (assuming the documentation is in git; I haven't looked)

I'm not really a design person, so I don't have strong opinions beyond that.

>  It might also make sense to change the existing tag-based feeds to reflect updates and not just new works, which might meet your needs for non-email-based tracking.

Sadly, I don't think the work I want to follow has tags.

I've gone ahead and made the changes you suggested in the hopes that it'll help.  I also rebased onto master.  As usual, the tests still don't work for me locally, so hopefully Travis will pass; if not, I'll fix things up (probably tomorrow).

I haven't added any new tests yet; I'll wait to see how the existing ones do, and whether you all are interested in the PR overall.

**Comment by novalis** [comment]
The existing tests seem to be flaky; they fail for me on master.  Since my new test passes, I figure I probably didn't break anything else.  So, I guess I'll just wait to hear back.

**Comment by novalis** [comment]
Just wanted to ping on this to see if there's been any progress on the discussion.  Would you prefer that I implement the link tag solution? 

Thanks.

**Comment by sarken** [comment]
Hi!

Unfortunately, we didn't have time to take a thorough look at this in our last committee meeting. We'll be sure to leave you a note when we make our decision, though!

In the meantime, I took a look at the test failures, and I'm afraid they weren't our known flaky tests, but actual failures: the changes to `published_at` are causing a "publication date can't be in the future" error, preventing new works from being posted. Fixing it can definitely wait until we've decide whether we're interested in this feature, but I wanted to let you know just the same.


Sarken
Archive development team

**Comment by novalis** [comment]
Thanks.  Do you have a list of known-flaky tests, or shall I just look at them individually and try to guess?

**Comment by sarken** [comment]
It might not be all of them, but here's at least a partial list: https://github.com/otwcode/otwarchive/wiki/Automated-Testing#Known_Test_Failures

**Comment by novalis** [comment]
I decided to just drop that migration; the RSS feeds might be a bit weird in cases where people post a bunch of chapters at once, but it doesn't really matter.  The purpose is to follow stories, and it will work fine for that.

**Comment by novalis** [comment]
Ping?

**Comment by sarken** [comment]
Hi again, and sorry for the delay! We've had a lot of time-sensitive business to take care of recently, so it took us a while to discuss this. We reviewed our feature requests and concluded that this isn't something we've had much demand for, so we'd prefer not to merge it and take on the overhead of maintaining it.

However, we're very grateful for your interest in AO3! If you'd like to contribute, our list of approved issues is available on JIRA: https://otwarchive.atlassian.net/projects/AO3/issues

**Comment by grey280** [comment]
Future reader here: I'm using a flaky RSS-bridge instance to replicate some of this functionality, and I have no idea how one is supposed to report User Demand For This Feature, so I'm just putting a comment here to say, please let me use RSS feeds for more things! Even just "author `:id` posted something new" would cover the vast majority of my use cases.

**Comment by Bilka2** [comment]
Thanks for the feedback! Our Support team monitors feature requests for us, so in the future, you can send them to [AO3 Support](https://archiveofourown.org/support).
